cpu/sam0_common/periph/uart.c

```diff
@@ -317,6 +317,44 @@ void uart_poweroff(uart_t uart)
     sercom_clk_dis(dev(uart));
 }
 
+#ifdef MODULE_PERIPH_UART_COLLISION
+bool uart_collision_detected(uart_t uart)
+{
+    bool collision = dev(uart)->STATUS.bit.COLL;
+    dev(uart)->STATUS.reg = SERCOM_USART_STATUS_COLL;
+    return collision;
+}
+
+void uart_collision_detect_enable(uart_t uart)
+{
+    /* clear stale collision flag */
+    dev(uart)->STATUS.reg = SERCOM_USART_STATUS_COLL;
+
+    /* enable collision detection */
+    dev(uart)->CTRLB.bit.COLDEN = 1;
+
+    /* disable RX interrupt */
+    dev(uart)->INTENCLR.bit.RXC = 1;
+}
+
+void uart_collision_detect_disable(uart_t uart)
+{
+    uint32_t ctrlb = dev(uart)->CTRLB.reg;
```

Comment:
You are ignoring the read-back bytes from the loopback of TX but at this point the receive buffer might not be empty. You could assert that it is empty by

```C
    /* clear readback bytes from receive buffer */
    while (dev(uart)->INTFLAG.bit.RXC) {
        (dev(uart)->DATA.reg);
    }
```

Also you might want to assert that CTRLB is synchronized at this point.

```C
    while (dev(uart)->SYNCBUSY.bit.CTRLB) { }
```